src/globalvars.cpp

```diff
@@ -76,18 +125,137 @@ void update_global_vars(const struct bpf_object *bpf_object,
   }
 
   // Update the values for the global vars (using the above offsets)
-  for (auto [name, offset] : vars_and_offsets) {
-    if (offset < 0) {
-      LOG(BUG) << "Global variable has not been added to the BPF code "
-                  "(codegen_llvm)";
-    }
-
+  for (const auto [global_var, offset] : vars_and_offsets) {
     int64_t *var = reinterpret_cast<int64_t *>(global_vars_buf + offset);
 
-    if (name == NUM_CPUS) {
+    if (global_var == GlobalVar::NUM_CPUS) {
       *var = bpftrace.ncpus_;
+    } else if (global_var == GlobalVar::MAX_CPU_ID) {
+      *var = bpftrace.max_cpu_id_;
+    }
+  }
+}
+
+static void update_global_vars_custom_rw_section(
+    const struct bpf_object *bpf_object,
+    const std::string &section_name,
+    struct bpf_map *global_vars_map,
+    const std::unordered_set<GlobalVar> &needed_global_vars,
+    BPFtrace &bpftrace)
+{
+  if (needed_global_vars.size() > 1) {
+    LOG(BUG) << "Multiple read-write global variables are in same section "
+             << section_name;
+  }
+  auto global_var = *needed_global_vars.begin();
+
+  size_t actual_size;
+  auto buf = bpf_map__initial_value(global_vars_map, &actual_size);
+  if (!buf) {
+    LOG(BUG) << "Failed to get size for section " << section_name
+             << " before resizing";
+  }
+  if (actual_size == 0) {
+    LOG(BUG) << "Section " << section_name << " has size of 0 ";
+  }
+
+  auto desired_size = (bpftrace.max_cpu_id_ + 1) * actual_size;
+  auto err = bpf_map__set_value_size(global_vars_map, desired_size);
+  if (err != 0) {
+    throw bpftrace::FatalUserException("Failed to set size to " +
+                                       std::to_string(desired_size) +
+                                       " for section " + section_name);
+  }
+
+  buf = bpf_map__initial_value(global_vars_map, &actual_size);
+  if (!buf) {
+    LOG(BUG) << "Failed to get size for section " << section_name
+             << " after resizing";
+  }
+  if (actual_size != desired_size) {
+    throw bpftrace::FatalUserException(
```

Comment:
Remind me, why is this a user error instead of a bug?

Reply:
You could argue it both ways:

User's fault because they don't have enough memory or on a computer with too many CPUs

vs. 

Our bug because we designed global var system

I don't feel strongly, I can see why this could be our bug but @danobi suggested a user facing error

Reply:
IMO anything that can fail which has an outside element should be handled gracefully. BUG (ie coredump) should be things firmly inside our control.